### PR TITLE
GPS: Improve the SBF detection and configuration loop

### PR DESCRIPTION
**Pull request by WickedShell**
This reworks the SBF configuration loop to not send any configuration until we get the prompt from the GPS unit which indicates that it is ready to handle data. This also allows us to shorten the delay between sending each one of these commands and leads to a faster configuration loop. I tested this by hard resetting a the GPS receiver. Without this the HitecMosaic for example would fail to ever redetect the GPS.

At the same time we can now pass an additional set of hwdef specific commands to the SBF configuration loop, which allows us to do things like ensure the GPS SD card logging is setup appropriately. I've included an example of this in the HitecMosaic hwdef. (Except for the log name I think these are actually the better ways to be running your GPS logging).

@amilcarlucas I think this removes the need for #8253 if you have a setup to test with that would be great. One potential change we might need is to unset the `readyForCommand` flag whenever we inject data, but I don't have a good setup at the moment to test with.

### CODE REVIEW

File: libraries/AP_GPS/AP_GPS_SBF.cpp [hunk 1]
```diff
@@ -155,6 +156,33 @@ AP_GPS_SBF::parse(uint8_t temp)
...
+            } else {
+                // attempt to detect command prompt
+                portIdentifier[portLength++] = (char)temp;
+                bool foundPossiblePort = false;
+                for (const char *portId : portIdentifiers) {
+                    if (strncmp(portId, portIdentifier, MIN(portLength, 3)) == 0) {
+                        // we found one of the COM/USB/IP related ports
+                        if (portLength == 4) {
+                            // validate that we have an ascii number
+                            if ((((char)temp) >= '0') && (((char)temp) <= '9')) {
```
Comment (tridge) [inline libraries/AP_GPS/AP_GPS_SBF.cpp:168]: isdigit() would be better

**Review by tridge — CHANGES_REQUESTED** [review]
could this get in a state where it will never get the prompt?

File: libraries/AP_GPS/AP_GPS_SBF.cpp (hunk 1 above)
Comment (WickedShell) [reply libraries/AP_GPS/AP_GPS_SBF.cpp:168]: Good call, swapped it.

**Comment by WickedShell** [comment]
> could this get in a state where it will never get the prompt?

To the best of my knowledge we should always get the prompt after sending the port enable string. The manual states this:

> Sometimes a connection is not configured to accept user commands, for example because it is put into differential correction input mode. A way to force a connection to accept commands is to send a succession of ten “S” characters to that connection and then to press the enter key (SSSSSSSSSS<CR>).


From my understanding of this we should always get the prompt afterwards. The manual does state several times that we must wait for the GPS to be ready for commands before we send it:

> The prompt indicates the termination of the processing of a given command. When sending multiple commands to the receiver, it is necessary to wait for the prompt between each command.

We weren't previously doing this, which I think is the root cause of why we would spam commands and they would not have any effect.

**Review by tridge — APPROVED** [review]

**Comment by amilcarlucas** [comment]
Thanks @WickedShell. I will test this and let you know how it goes.

**Comment by peterbarker** [comment]
@amilcarlucas nudge

**Comment by WickedShell** [comment]
Merging as testing appears to be externally stalled and is working well in my testing. If this causes any regressions in the SBF driver please ping me.

**Pull request merged**